src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderCF.java

```diff
@@ -0,0 +1,99 @@
+package graphql.execution.instrumentation.dataloader;
+
+import graphql.ExperimentalApi;
+import graphql.Internal;
+import graphql.execution.DataLoaderDispatchStrategy;
+import graphql.execution.ExecutionContext;
+import graphql.schema.DataFetchingEnvironment;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Supplier;
+
+import static graphql.execution.Execution.EXECUTION_CONTEXT_KEY;
+
+@Internal
+public class DataLoaderCF<T> extends CompletableFuture<T> {
+    final DataFetchingEnvironment dfe;
+    final String dataLoaderName;
+    final Object key;
+    final CompletableFuture<Object> dataLoaderCF;
+
+    volatile CountDownLatch latch;
+
+    public DataLoaderCF(DataFetchingEnvironment dfe, String dataLoaderName, Object key) {
+        this.dfe = dfe;
+        this.dataLoaderName = dataLoaderName;
+        this.key = key;
+        if (dataLoaderName != null) {
+            dataLoaderCF = dfe.getDataLoaderRegistry().getDataLoader(dataLoaderName).load(key);
+            dataLoaderCF.whenComplete((value, throwable) -> {
+                System.out.println("underlying DataLoader completed");
+                if (throwable != null) {
+                    completeExceptionally(throwable);
+                } else {
+                    complete((T) value);
+                }
+                // post completion hook
+                if (latch != null) {
+                    latch.countDown();
```

Comment:
Maybe rather than a latch -we have a `Consumer< DataLoaderCF>` and we call `consumer.accept(this)`

That way the DataLoaderCF can have different "implementations" of what happens when its done.  Its dumber and hence can evolve a little better

I mean in theory If the engine sees the `DataLoaderCF` it could put a `.whenComplete()` on it and do this countdown latch twizzling itself.  Therefore you dont need this to be mutavle.

I think that would be better

Reply:
Not sure I understand it correctly but the problem with when complete is that it fires first ... and you can't add a dependent CF to the beginning of the stack of dependents. This is why the latch is called after the complete call.